core/src/main/java/org/apache/calcite/rel/metadata/AbstractRelMetadataQuery.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.metadata;
+
+import org.apache.calcite.rel.RelNode;
+
+import java.lang.reflect.Proxy;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+/**
+ * Base class for the RelMetadataQuery that uses the metadata handler class
```

Comment:
Not sure why we need this abstract class. The custom meta data query class would derive from RelMetadataQuery.

Reply:
I still feel the abstract/base class is not needed. If we really want to re-configure the build-in meta handlers, we can always refactor in the future.

Reply:
The base class is not absolutely essential. But it separates the mechanism from the handler-specific stuff. The mechanism and handler-specific stuff are different kinds of code, and indication that they should be in different classes.

It will make life better for the derived class because the class that has the built-in handlers isn't getting any special privileges.

We should consider making the generated code use the base class.

Does anyone think the base class should be an embedded object, not a base class? That's an option to consider.